src/client/Microsoft.Identity.Client/Platforms/Android/AndroidBroker.cs

```diff
@@ -0,0 +1,194 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Android.Accounts;
+using Android.App;
+using Android.Content;
+using AndroidNative = Android;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Internal.Broker;
+using Microsoft.Identity.Client.OAuth2;
+using Microsoft.Identity.Client.UI;
+using Microsoft.Identity.Client.Utils;
+using System.Globalization;
+using Microsoft.Identity.Json.Linq;
+using Microsoft.Identity.Json;
+using Android.OS;
+
+namespace Microsoft.Identity.Client.Platforms.Android
+{
+    [AndroidNative.Runtime.Preserve(AllMembers = true)]
+    internal class AndroidBroker : IBroker
+    {
+        private static SemaphoreSlim readyForResponse = null;
+        private static MsalTokenResponse resultEx = null;
+        private readonly AndroidBrokerHelper _brokerHelper;
+        private readonly ICoreLogger _logger;
+        private Activity _activity;
+        //TODO: Temp variable to hold data not requrned from the broker due to a bug. Will be removed.
+        private static Dictionary<string, string> _brokerPayload;
+
+        public AndroidBroker(ICoreLogger logger)
+        {
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _brokerHelper = new AndroidBrokerHelper(Application.Context, logger);
+        }
+
+        public bool CanInvokeBroker(CoreUIParent uiParent)
+        {
+            bool canInvoke = _brokerHelper.CanSwitchToBroker();
+            _logger.Verbose("Can invoke broker? " + canInvoke);
+            _activity = uiParent.Activity;
+            return canInvoke;
+        }
+
+        public async Task<MsalTokenResponse> AcquireTokenUsingBrokerAsync(Dictionary<string, string> brokerPayload)
+        {
+            resultEx = null;
+            readyForResponse = new SemaphoreSlim(0);
+            _brokerPayload = brokerPayload;
+            //Need to disable warning for non awaited async call.
+#pragma warning disable 4014
+            try
+            {
+                Task.Run(() => AcquireTokenInternalAsync(brokerPayload));
+            }
+            catch (MsalException Ex)
+            {
+                _logger.Error("Broker Operation Failed to complete.");
+                throw Ex;
```

Comment:
Prefer to use `throw` instead of `throw ex;` - this is because if you do `throw` the exception keeps its stack trace. If you do `throw ex` the stack trace will simply show line 65